pkg/loop/client.go

```diff
@@ -0,0 +1,154 @@
+
+var _ grpc.ClientConnInterface = (*clientConn)(nil)
+
+// newClientFn returns a new client connection id to dial, and a set of resource dependencies to close.
+type newClientFn func(context.Context) (id uint32, deps resources, err error)
+
+// clientConn is a [grpc.ClientConnInterface] backed by a [*grpc.ClientConn] which can be recreated and swapped out
+// via the provided [newClientFn].
+type clientConn struct {
+	*lggrBroker
+	newClient newClientFn
+	name      string
+
+	mu   sync.RWMutex
+	deps resources
+	cc   *grpc.ClientConn
+}
+
+func (lb *lggrBroker) newClientConn(name string, newClient newClientFn) *clientConn {
+	return &clientConn{
+		lggrBroker: lb.named(name),
+		newClient:  newClient,
+		name:       name,
+	}
+}
+
+func (c *clientConn) Invoke(ctx context.Context, method string, args interface{}, reply interface{}, opts ...grpc.CallOption) error {
+	c.mu.RLock()
+	cc := c.cc
+	c.mu.RUnlock()
+
+	if cc == nil {
+		cc = c.refresh(ctx, nil)
+	}
+	for cc != nil {
+		err := cc.Invoke(ctx, method, args, reply, opts...)
+		if err == nil || !isErrTerminal(err) {
+			return err
+		}
+		cc = c.refresh(ctx, cc)
+	}
+	return context.Cause(ctx)
+}
+
+func (c *clientConn) NewStream(ctx context.Context, desc *grpc.StreamDesc, method string, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+	c.mu.RLock()
+	cc := c.cc
+	c.mu.RUnlock()
+
+	if cc == nil {
+		cc = c.refresh(ctx, nil)
+	}
+	for cc != nil {
+		s, err := cc.NewStream(ctx, desc, method, opts...)
+		if err == nil || !isErrTerminal(err) {
+			return s, err
+		}
+		cc = c.refresh(ctx, cc)
+	}
+	return nil, context.Cause(ctx)
+}
+
+func (c *clientConn) refresh(ctx context.Context, orig *grpc.ClientConn) *grpc.ClientConn {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.cc != orig {
+		return c.cc
+	}
+	if c.cc != nil {
+		if err := c.cc.Close(); err != nil {
+			c.lggr.Errorw("Client close failed", "err", err)
+		}
+		c.closeAll(c.deps...)
+	}
+
+	try := func() bool {
+		c.lggr.Debug("Client refresh")
+		id, deps, err := c.newClient(ctx)
+		if err != nil {
+			c.lggr.Errorw("Client refresh failed", "err", err)
+			c.closeAll(deps...)
+			return false
+		}
+		c.deps = deps
+
+		lggr := logger.With(c.lggr, "id", id)
+		lggr.Debug("Client dial")
+		c.cc, err = c.broker.Dial(id)
+		if err != nil {
+			err = ErrConnDial{Name: c.name, ID: id, Err: err}
+			lggr.Errorw("Client dial failed", "err", err)
+			c.closeAll(c.deps...)
+			return false
+		}
+		return true
+	}
+
+	const maxWait = 5 * time.Second
+	for wait := 100 * time.Millisecond; !try(); wait *= 2 {
+		if wait > maxWait {
+			wait = maxWait
+		}
+		c.lggr.Infow("Waiting to refresh", "wait", wait)
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(wait):
+		}
+	}
+
+	return c.cc
+}
+
+func isErrTerminal(err error) bool {
+	switch status.Code(err) {
+	case codes.Unavailable, codes.Canceled:
+		return true
+	case codes.OK, codes.Unknown, codes.InvalidArgument, codes.DeadlineExceeded, codes.NotFound, codes.AlreadyExists,
```

Comment:
why bother making these explicit? would be easier to read if only the small set of terminal errs is explicit

Reply:
what are the rules for judging whether an err code is terminal? what happens if we wrongly judge one?

Reply:
The linter insists on exhaustively listing types like this.

Reply:
If we are wrong then we return it to the caller instead of refreshing the connection.

Reply:
> re:linter doesn't a default case with return false work to avoid explicit false case?

Works now (and that's what I wrote originally), but is brittle since it pre-classifies future types I suppose. If a new type is introduce, we would be forced to classify it explicitly.
https://golangci-lint.run/usage/linters/#exhaustive
https://pkg.go.dev/github.com/nishanths/exhaustive

> re: bad judgement - does that imply that we could get stuck with a bad connection indefinitely?

Yes, and the opposite mistake would needlessly recreate clients for non-terminal cases.

Reply:
yeah, both are bad.

getting stuck with a bad connection seems scarier.

why would something like Unimplemented not be terminal?

Reply:
`Unimplemented` is an actual response from a live server - just that it doesn't serve that method.